idioms/fallible-closure.md

```diff
@@ -0,0 +1,35 @@
+# Fallible closure inside infallible function
+
+## Description
+
+Build a fallible closure inside of an apparently infallible function.
+
+## Example
+
+```rust
+use std::path::Path;
+
+// this function hides its failure states
+fn file_stem(path: &Path) -> &str {
+    // this closure is fallible, so it can make use of `?`
+    // fn() -> Option<&str>
+    let inner = || Some(path.file_stem()?.to_str()?);
```

Comment:
I believe this example is not making full use of `?`, it could even have `TryInto` IIRC.

Reply:
It could be written :

```rust
let inner =  || path.file_stem()?.to_str();
```

but that wouldn't show much

Reply:
I do think that is better. I don't know why we need to have another layer for that.